doc/developer/design/20231002_cluster_vision.md

```diff
@@ -0,0 +1,133 @@
+# Cluster UX Long Term Vision
+
+- Associated: [Epic](https://github.com/MaterializeInc/materialize/issues/22120)
+
+## The Problem
+We need a documented vision for the cluster UX in the long term which covers both
+the "end state" goal as well as the short and medium states in order to:
+* Ensure alignment in the future that we are working toward
+* Make product prioritization decisions around cluster work
+* Make folks more comfortable accepting intermediate states that aren't ideal in service of a greater goal
+* Come up with a narrative for customers on what to expect around cluster management
+
+Epic: https://github.com/MaterializeInc/materialize/issues/22120
+
+## Success Criteria
+Primarily, a merged design doc that is reviewed and approved by EPD leadership,
+and is socialized to GTM.
+
+Secondarily, a roadmap for cluster work for the next quarter.
+
+Qualitatively, positive feedback from EPD leadership and GTM folks that they
+have clarity on the vision and roadmap, and the reasoning behind those decisions.
+
+## Out of Scope
+Designing the actual cluster API changes themselves, or proposing implementation details.
+
+## Solution Proposal
+The objectives we are striving for with the cluster UX:
+* Easy to use and manage
+* Maximize resource efficiency/minimize unused resource cost
+* Enable fault tolerance/use-case isolation
+
+### Declarative vs Imperative
+We should move toward a declarative API for managing clusters, where:
+
+Declarative is like `CREATE CLUSTER` with managed replicas and \
+Imperative is like `CREATE/DROP CLUSTER REPLICA`.
+
+This means deprecating manual cluster replica management. \
+We believe this is easier to use and manage.
+
+The primary work item for this is **graceful reconfiguration**. At the moment, a change in size causes downtime until the new replicas are hydrated. As such, customers still want the flexibility to create their own replicas for graceful resizing. We can avoid this by leaving a subset of the original replicas around until the new replicas are hydrated. \
```

Comment:
I'm increasingly wondering if we _don't_ need graceful reconfiguration. At least, if you buy into the theory that blue/green is you do things gracefully ... then perhaps `ALTER ... SET CLUSTER` _should_ be a dangerous forcing operation that puts things into action immediately, even if doing so causes downtime.

Reply:
Ah kind of like what we're thinking for indexes, where if you want to in-place drop an index that will incur downtime, and if you want to gracefully drop one we suggest blue/green?

My only hesitation there is that forcing users to do blue/green is more work and conceptual overhead for the user than graceful reconfiguration. From a UX standpoint, graceful reconfiguration is effectively an abstraction on top blue/green, that manages spinning up the new resource and cutting over for them.
Even with blue/green, we'd need (1) to give users an easy way to detect that rehydration is complete so that they can execute the cutover themselves. We just wouldn't need (2), the triggering mechanism.

Also I was thinking the syntax here would be `ALTER CLUSTER ... SET (SIZE = <>)`, which spins up different sized replicas and cuts over to those replicas. As it stands now, adding another replica to an existing cluster doesn't require users to drop and recreate all downstream dependencies. My understanding is that blue/green requires recreating a whole second version of the stack, including all objects and clusters, which is a much heavier lift.

I think we should strive for graceful reconfiguration as an end state for resizing, with blue/green as an intermediary state.
For moving objects between clusters, I'm fine with that continuing to require blue/green.

Reply:
> Even with blue/green, we'd need (1) to give users an easy way to detect that rehydration is complete so that they can execute the cutover themselves.

Yep, definitely! But I think we're much closer to being able to surface that rehydration complete signal than we are to being able to take action off of it directly inside of Materialize. 

It's a good point that moving objects between clusters and scaling up a cluster are meaningfully different operations. Moving an object between clusters is likely part of your development workflow—e.g., maybe a big refactoring of how you map objects to clusters that you'll test extensively with a blue/green setup. But scaling up/down a cluster is likely something that you do live on the system (e.g., in response to load) and you both don't want that to cause downtime but it's overkill to reach for a `dbt`-managed blue/green deployment just to scale up/down.